Various changes for evolutions

nilsdeppe opened the pull request:
## Proposed changes

Make various changes that will be necessary for evolutions. I'm splitting the SW evolution PR into several smaller PRs so they're easier to deal with. This is the first of the series of PRs

### Types of changes:

- [x] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."


### Further comments

This will conflict with PRs that change DomainCreators because the base class is changed in this PR. Specifically, #336 will conflict with these changes.

closes #274

nilsdeppe committed:
Add register_derived_classes_with_charm to Parallel namespace

osheamonn commented on src/Domain/DomainCreators/Interval.cpp:
```diff
@@ -4,6 +4,7 @@
 #include "Domain/DomainCreators/Interval.hpp"
 
 #include <array>
+#include <charm.h>
```
Is this include needed? It's not added in any other domain creators.

nilsdeppe replied:
Unintentional, will remove :)

osheamonn commented on src/DataStructures/DataBox.hpp:
```diff
@@ -780,14 +785,43 @@ SPECTRE_ALWAYS_INLINE constexpr void add_variables_compute_tags_to_box(
+    typelist<Ts...> /*meta*/) {
+  const auto helper = [&data](auto tag_v) {
+    using tag = decltype(tag_v);
+    auto& vars = get<tag>(data.template get<VariablesTag>().mutate());
```
Does the free `get` function work with TaggedDeferredTuple?

nilsdeppe replied:
I have two ways of interpreting your question, so I'll answer both cases:

1. TaggedDeferredTuple does not currently have a free get function because when I brought over the modified TaggedTuple I did not want to hunt through and replace all the `.template get` uses inside DataBox. This can be changed, but I think an issue should be filed and it should be done in a separate PR.

2. If you're confused about the outer `get` call, it is not on the TaggedDeferredTuple.

Let me know if this answers your question.

osheamonn replied:
Yeah the first interpretation is what I was intending. Sounds reasonable.

nilsdeppe replied:
Alright, made the issue: #346

kidder commented on src/DataStructures/DataBox.hpp:
```diff
@@ -553,9 +553,9 @@ struct create_dependency_graph {
-  using sub_tree =
-      tmpl::fold<typename Callee::argument_tags, List,
-                 create_dependency_graph<Callee, tmpl::_element, tmpl::_state>>;
+  using sub_tree = tmpl::fold<
```
Please add a better commit message, as we had no initial idea what this commit is doing

nilsdeppe replied:
Looking over this, I should have made 3 separate commits, I will refactor it so that it's clear what's happening. The commits will be:

1. compute_databox_type metafunction
2. fix mutate bug (resetting of Variables-compute items did not work correctly, leading to confusing compiler errors at best)
3. mutate_apply function

kidder replied:
👍

kidder commented on src/DataStructures/DataBox.hpp:
```diff
@@ -1040,11 +1135,11 @@ constexpr DataBox<TagsList<Tags...>>::DataBox(
       "the function creating the new DataBox.");
   // Create dependency graph between compute items and items being reset
-  using edge_list =
+  using local_edge_list =
```
we assume the prepend of local_ is to avoid name conflicts?  I vaguely remember not liking local for this purpose in an earlier PR, but we don't remember why or what we had local changed to

nilsdeppe replied:
I remember this being an issue for template parameters, and I think the issue was changing `LArgs` -> `LocalArgs`. In many cases they were also `List` -> `Pack` so then the `Local` was no longer necessary. How about `kept_compute_item_edge_list`? It's a bit long but that is exactly what it is

kidder replied:
👍

kidder commented on src/DataStructures/DataBox.hpp:
```diff
@@ -1381,6 +1476,39 @@ inline constexpr auto apply(F f, const DataBox<BoxTags...>& box,
+}
+}  // namespace databox_detail
+
+template <typename ReturnTags, typename ArgumentTags, typename F,
```
Please document this public function 😈

nilsdeppe commented:
I've done a few things after the comments:

1. Added a `CREATE_IS_CALLABLE` macro, so this PR closes #274 
2. Split the DataBox commit into 4 commits.
  - Add mutate_apply function
  - Fix a bug in db::mutate that resulted in Variables compute items not resetting correctly
  - Add a metafunction to compute the type of a DataBox given tags
  - Add an error test for get_lazy

nilsdeppe committed:
DomainCreators return vector of initial refinement/extents

Since we cannot serialize DomainCreators getting the initial
refinement level or extents is not useful for a single block. Instead
a vector of refinement levels or extents is returned by the
DomainCreator. The integer ID of the Block must match with the
location of that Block's entries in the vectors.

nilsdeppe committed:
Add db::mutate_apply function

Function allows the invoked function to mutate items in the DataBox.

nilsdeppe committed:
Neighbors.hpp doesn't depend on Direction

nilsdeppe committed:
Include Main and ConstGlobalCache in CharmMain

nilsdeppe committed:
Use CREATE_IS_CALLABLE in Algorithm

nilsdeppe committed:
Add macro CREATE_IS_CALLABLE to generate type traits

The macro generates type traits that can be used to check if a class
has a method that can be called with arguments of specified types.

nilsdeppe committed:
Add test of operator<< for Index

nilsdeppe committed:
Add error test for DataBox's get_lazy function

nilsdeppe committed:
Fix bug in db::mutate

1. Compute items dependent on the result of a Variables compute item were
not properly being reset.

2. Bug in metaprogram resulted in compilation errors

nilsdeppe committed:
Add metafunction to compute the type of a DataBox given tags

nilsdeppe committed:
Add tags for initial time and initial delta T

nilsdeppe commented:
I pushed updates that fix several other issues with Variables and resetting compute items correctly. As a result a bunch of code was replaced by a newer implementation, so I removed that as well. The only commits affected by this were `Fix bug in db::mutate` and `Add db::mutate_apply`

kidder commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -780,6 +780,47 @@ template <typename T, typename... Args>
+ * \brief Generate a type trait to check if a class has a member function that
+ * can be invoked with arguments of type `TArgs...`
+ *
+ * The usage of the type trait is identical to the usage to the usage the of
```
what were you ...?

kidder commented on src/DataStructures/DataBox.hpp:
```diff
@@ -541,6 +535,25 @@ struct extract_dependent_items<
+};
+
+// Given a typelist List, returns a new typelist with all the Variables Tags'
+// members
```
delete members

kidder commented on tests/Unit/DataStructures/Test_DataBox.cpp:
```diff
@@ -527,16 +527,183 @@ struct VectorTag : db::DataBoxTag {
+  CHECK(db::get<test_databox_tags::VectorTag2>(box) ==
+        (tnsr::I<DataVector, 3>(DataVector(2, 2.))));
+
+  auto box2 = db::create_from<
```
use db::AddTags, etc.

kidder replied:
do you test resetting a tensor in a variables for a variables compute item?

wthrowe replied:
ping (on testing variables compute item after resetting a tensor)

nilsdeppe replied:
This is tested above, line 736

nilsdeppe replied:
Err, sorry 680 and 701 & 703

nilsdeppe replied:
I'll add another check here to be sure it works in both cases

wthrowe replied:
I want you to check that `MultiplyVariablesByTwo` resets correctly.  You have already put this in the DataBox but never use it.

nilsdeppe replied:
I do use it, line 701 and 703, but I agree that it's good to have an additional check so I'm adding:

```cpp
  {
    const auto& vars = db::get<test_databox_tags::MultiplyVariablesByTwo>(box);
    CHECK(get<test_databox_tags::ScalarTag4>(vars) ==
          Scalar<DataVector>(DataVector(2, 6.)));
    CHECK(get<test_databox_tags::VectorTag4>(vars) ==
          (tnsr::I<DataVector, 3>(DataVector(2, 6.))));
  }
```

Is this what you're looking for?

I'm  adding this to both the db::mutate test and create_from test

kidder commented on src/DataStructures/DataBox.hpp:
```diff
@@ -1465,6 +1466,99 @@ inline constexpr auto apply(F f, const DataBox<BoxTags...>& box,
+inline constexpr auto mutate_apply(F f, db::DataBox<BoxTags...>& box,
+                                   tmpl::list<ReturnTags...> /*meta*/,
+                                   tmpl::list<ArgumentTags...> /*meta*/,
+                                   Args&&... args) {
```
be consistent with noexcept

kidder commented on src/DataStructures/DataBox.hpp:
```diff
@@ -1465,6 +1466,99 @@ inline constexpr auto apply(F f, const DataBox<BoxTags...>& box,
+ * \snippet Test_DataBox.cpp mutate_apply_example
+ *
+ * An example of a class with a static `apply` function
+ *
```
if possible remove blank line

kidder commented on src/DataStructures/DataBox.hpp:
```diff
@@ -1465,6 +1466,99 @@ inline constexpr auto apply(F f, const DataBox<BoxTags...>& box,
+ * \tparam MutateTags typelist of Tags to mutate
+ * \tparam ArgumentTags typelist of additional items to retrieve from the
+ * DataBox
+ * \param f the function to apply \param box the DataBox out of which to
```
unsquish

wthrowe replied:
ping

nilsdeppe replied:
*angrily shakes fist at clang-format*

kidder commented on src/DataStructures/DataBox.hpp:
```diff
@@ -550,7 +550,6 @@ struct closure_extract_dependent_items<true> {
 };
 
 // Given a typelist List, returns a new typelist with all the Variables Tags'
-// members
```
remove the ' after Tags

kidder reviewed (changes requested):
I'm happy for you to squash

nilsdeppe commented:
Okay, rebased on develop and squashed!

nilsdeppe commented on src/DataStructures/DataBox.hpp:
```diff
@@ -1381,6 +1493,117 @@ inline constexpr auto apply(F f, const DataBox<BoxTags...>& box,
+ * \tparam MutateTags typelist of Tags to mutate
+ * \tparam ArgumentTags typelist of additional items to retrieve from the
+ * DataBox
+ * \param f the function to apply
```
Okay, finally it didn't change it back -_-

nilsdeppe committed:
Add make_list metafunction for debugging

nilsdeppe commented:
I addressed @wthrowe 's request, added a static_assert when trying to mutate a member of a Variables in `create_from`, and added a commit for the `make_list` metafunction, which is useful for debugging. The doxygen comment for `make_list` should be useful, so please request changes if it isn't.

kidder reviewed (changes requested):
rename the file TmplDebugging.hpp

fmahebert commented on src/Utilities/TypeDisplayer.hpp:
```diff
@@ -22,6 +24,74 @@
+ * using variables_tags_from_single_tags2 = tmpl::transform<
+ *     extracted_from_variables,
+ *     tmpl::bind<tmpl::transform, tmpl::pin<mutated_tags_list>,
+ *                tmpl::bind<make_list, tmpl::_1, tmpl::parent<tmpl::_1>>>>;
```
previous text says to use `tmpl::make_list`

nilsdeppe replied:
Thanks! Fixed!

nilsdeppe committed:
Have Travis check for TmplDebugging include

kidder commented on tools/CheckFiles.sh:
```diff
@@ -22,8 +22,21 @@ iostream_test() {
+    is_c++ "$1" && grep -q '#include "Utilities/TmplDebugging.hpp"' "$1"
+}
+tmpl_debugging_report() {
+    echo "Found iostream header:"
```
RIP
